File: numguess.rb
class NumGuess
	DEFAULT_NAME = "Player"
	MIN_LIMIT = 10

	@name = DEFAULT_NAME
	@limit = MIN_LIMIT
	@num = 0
	@tries = 0
	@max_tries = 0
	@correct = false
	
	def name=(name = DEFAULT_NAME)
		@name = name.empty? ? DEFAULT_NAME : name
	end
	
	def limit=(limit)
		@limit = limit < MIN_LIMIT ? MIN_LIMIT : limit
		@max_tries = Math.log(@limit, 2).floor + 1
	end
	
	def start_game
		@correct = false
		@tries = 0
		@num = rand(1..@limit)
	end
	
	def evaluate_guess(guess)
		return "That's just plain wrong." if guess.to_i.to_s != guess

		guess = guess.to_i
		return "Out of range." unless (1..@limit) === guess
		
		@tries += 1
		case guess <=> @num
			when -1
				return "Too low!"
			when 1
				return "Too high!"
			else
				@correct = true
				return ""
		end
	end
	
	def custom_message
		return "You're one lucky bastard!"              if @tries == 1
		return "You are the master of this game!"       if @tries < @max_tries
		return "You are a machine!"                     if @tries == @max_tries
		return "Very good result!"                      if @tries < @max_tries * 1.1
		return "I find your lack of skill disturbing!"  if @tries > @limit
		return "Try harder, you can do better!"
	end
	
	def run
		print "Welcome to NumGuess Ruby version!\n\n"
		print "Enter your name: "
		self.name = gets.chomp

		print "\nWelcome #{@name}, enter upper limit: "
		self.limit = gets.chomp.to_i

		begin
			print "\nGuess my number between 1 and #{@limit}!\n"
			start_game
			
			begin
				print "\nGuess: "
				print evaluate_guess(gets.chomp)
			end until @correct
			
			print "\nWell done #{@name}, you guessed my number from #{@tries} %s!\n" %
				(@tries == 1 ? "try" : "tries")
			
			print custom_message + "\n"
			print "Play again [Y/N]? "
		end until gets.chomp.downcase != "y"

		print "\nOkay, bye."
	end
end

NumGuess.new.run

File: numguess.R
##########################################################################################
##
## Use the "source" command in the R console to run this game! E.g.:
##
##   library(RCurl)
##   source(textConnection(
##     getURL('https://raw.githubusercontent.com/DarthJDG/NumGuess/master/numguess.R')))
##
##########################################################################################

#' Transform string into integer
#' @param x character vector
#' @return integer
int <- function(x)
    suppressWarnings(ifelse(as.numeric(x) == as.integer(x), as.integer(x), NA))
#' Cat sprintf results with ending line-break
#' @param fmt "character vector of format strings" to be passed to \code{sprintf}
#' @param ... values to be passed to \code{sprintf}
catsn <- function(fmt, ...)
    cat(sprintf(fmt, ...), '\n')

## say hello
catsn('Welcome to NumGuess R version!\n')

## get username
name <- readline('Enter your name: ')

## and set the default one if nothing was provided
if (name == '')
    name <- 'Player'

## get upper limit
limit <- readline(sprintf('\nWelcome %s, enter upper limit: ', name))

## validate limit
limit <- int(limit)
if (is.na(limit) || limit < 10)
    limit <- 10L

## maximum number of tries
max_tries <- floor(log2(limit)) + 1

## infinite loop
while (TRUE) {

    catsn('\nGuess my number between 1 and %d!\n', limit)

    ## generate a number between 1 and limit
    number <- sample(1:limit, 1)
    tries  <- 0

    ## internal loop for guessing
    while (TRUE) {

        ## ask for a number
        guess <- readline('Guess: ')

        ## validate guess
        guess <- int(guess)
        if (is.na(guess)) {
            catsn('That\'s just plain wrong.')
            next()
        }
        if (guess < 1 | guess > limit) {
            catsn('Out of range.')
            next()
        }

        ## check if we have a winner
        tries <- tries + 1
        if (number == guess)
            break()
        if (number < guess)
            catsn('Too high!')
        if (number > guess)
            catsn('Too low!')

    }

    ## congrats
    catsn(
        '\nWell done %s, you guessed my number from %d %s!',
        name,
        tries,
        ifelse(tries == 1, 'try', 'tries'))

    ## custom message
    cm <- ifelse(
        tries == 1,
        'You\'re one lucky bastard!',
        ifelse(
            tries < max_tries,
            'You are the master of this game!',
            ifelse(
                tries == max_tries,
                'You are a machine!',
                ifelse(
                    tries < max_tries * 1.1,
                    'Very good result!',
                    ifelse(
                        tries < limit,
                        'Try harder, you can do better!',
                        'I find your lack of skill disturbing!'
                    )
                )
            )
        )
    )
    catsn(cm)

    ## next round
    nr <- readline('Play again [y/N]? ')
    if (tolower(nr) != 'y')
        break()

}

## say goodbye
catsn('\nOkay, bye.')
